src/webservice/Router.cpp

```diff
@@ -0,0 +1,123 @@
+
+void Route::checkPath(const std::string &path) {
+    CHECK(!path.empty() && path[0] == '/') << "Path must start with '/'";
+}
+
+void Route::setPath(const std::string &path) {
+    auto loc = path.find(':');
+    if (loc != std::string::npos) {
+        createPattenRegex(path);
+        this->path_ = path;
+    } else {
+        // exact path
+        if (path.back() == '/') {
+            this->path_ = path.substr(0, path.length() - 1);
+        } else {
+            this->path_ = path;
+        }
+    }
+}
+
+std::unique_ptr<std::regex> Route::reToken_ =
+    std::make_unique<std::regex>(":([A-Za-z][A-Za-z_0-9]*)");
+
+void Route::createPattenRegex(const std::string &path) {
+    int pos = 0;
+    std::stringstream ss;
+    for (std::sregex_iterator next(path.begin(), path.end(), *reToken_), end; next != end; next++) {
+        std::string str = next->str();
+        if (!str.empty() && str.front() == ':') {
+            str = str.substr(1);
+        }
+        for (auto &group : groups_) {
+            if (str == group) {
+                LOG(FATAL) << "Cannot use identifier " << group
+                           << " more than once in pattern string";
+            }
+        }
+        groups_.emplace_back(str);
+        ss << path.substr(pos, next->position() - pos) << "([^/]+)";
+        pos = next->position() + next->str().length();
+    }
+    ss << path.substr(pos);
+    this->pattern_ = std::make_unique<std::regex>(ss.str());
+}
+
+bool Route::matches(proxygen::HTTPMethod method, const std::string &path) const {
+    if (method_ != method) {
+        return false;
+    }
+    std::string p = path;
+    if (p.size() > 1 && p.back() == '/') {
+        p = path.substr(0, path.length() - 1);
+    }
+    if (pattern_) {
+        return std::regex_search(p, *pattern_);
+    } else {
+        return path_ == p;
+    }
+}
+
+Router::~Router() {
+    while (head_ != nullptr) {
+        auto next = head_->next();
+        delete head_;
+        head_ = next;
+    }
+}
+
+proxygen::RequestHandler *Route::generateHandler(const std::string &path) const {
+    if (!pattern_) {
+        return generator_({});
+    }
+    std::smatch m;
+    std::regex_search(path, m, *pattern_);
+    PathParams params;
+    CHECK_EQ(groups_.size(), m.size() - 1UL) << "groups is not equal to matches";
+    for (std::size_t i = 0; i < groups_.size(); i++) {
+        params.emplace(groups_[i], m[i + 1]);
+    }
+    return generator_(std::move(params));
+}
+
+proxygen::RequestHandler *Router::dispatch(const proxygen::HTTPMessage *msg) const {
+    for (Route *r = head_; r != nullptr; r = r->next()) {
+        if (r->matches(msg->getMethod().value(), msg->getPath())) {
+            return r->generateHandler(msg->getPath());
+        }
+    }
+    return new NotFoundHandler();
+}
+
+void Router::append(Route *route) {
+    if (head_ != nullptr) {
+        // Add current route to tail
+        auto iter = head_;
+        while (iter->next() != nullptr) {
```

Comment:
Here begins from the head every time. what about adding a tail ptr?

Reply:
You are right. I consider that there are not too many routes for iterating, so I don't have a pointer to save the tail.  But I can fix it.